tests/unit/test_id_cli.py

```diff
@@ -0,0 +1,466 @@
+    workflows, _ = await parse_ids_async(*ids_in, constraint='tasks', src=True)
+    assert {
+        workflow_id: [detokenise(tokens) for tokens in tokens_list]
+        for workflow_id, tokens_list in workflows.items()
+    } == ids_out
+
+
+@pytest.mark.parametrize(
+    'ids_in,ids_out',
+    [
+        (('a//',), {'a': []}),
+        (
+            ('a//', 'b//', 'c//'),
+            {'a': [], 'b': [], 'c': []}
+        ),
+        (('a//i',), {'a': ['//i']}),
+        (('a//', '//i'), {'a': ['//i']}),
+        (
+            ('a//', '//i', '//j', '//k'),
+            {'a': ['//i', '//j', '//k']},
+        ),
+        (('a//', '//i', 'b//'), {'a': ['//i'], 'b': []}),
+    ]
+)
+async def test_parse_ids_mixed(ids_in, ids_out):
+    """It should parse mixed workflows & tasks."""
+    workflows, _ = await parse_ids_async(*ids_in, constraint='mixed')
+    assert {
+        workflow_id: [detokenise(tokens) for tokens in tokens_list]
+        for workflow_id, tokens_list in workflows.items()
+    } == ids_out
+
+
+@pytest.mark.parametrize(
+    'ids_in,ids_out',
+    [
+        (('./a',), {'a': []}),
+        (('./a', '//i'), {'a': ['//i']}),
+        (('./a', '//i', '//j', '//k'), {'a': ['//i', '//j', '//k']}),
+    ]
+)
+async def test_parse_ids_mixed_src(ids_in, ids_out, abc_src_dir):
+    """It should parse mixed workflows & tasks from src workflows."""
+    workflows, _ = await parse_ids_async(*ids_in, constraint='mixed', src=True)
+    assert {
+        workflow_id: [detokenise(tokens) for tokens in tokens_list]
+        for workflow_id, tokens_list in workflows.items()
+    } == ids_out
+
+
+@pytest.mark.parametrize(
+    'ids_in,errors',
+    [
+        (('a//',), False),
+        (('a//', 'b//'), False),
+        (('a//', 'b//', 'c//'), True),
+    ]
+)
+async def test_parse_ids_max_workflows(ids_in, errors):
+    """It should validate input against the max_workflows constraint."""
+    try:
+        await parse_ids_async(*ids_in, constraint='workflows', max_workflows=2)
+    except UserInputError:
+        if not errors:
+            raise
+    else:
+        if errors:
+            raise Exception('Should have raised UserInputError')
+
+
+@pytest.mark.parametrize(
+    'ids_in,errors',
+    [
+        (('a//', '//i'), False),
+        (('a//', '//i', '//j'), False),
+        (('a//', '//i', '//j', '//k'), True),
+    ]
+)
+async def test_parse_ids_max_tasks(ids_in, errors):
+    """It should validate input against the max_tasks constraint."""
+    try:
+        await parse_ids_async(*ids_in, constraint='tasks', max_tasks=2)
+    except UserInputError:
+        if not errors:
+            raise
+    else:
+        if errors:
+            raise Exception('Should have raised UserInputError')
+
+
+async def test_parse_ids_infer_run_name(tmp_run_dir):
+    """It should infer the run name for auto-numbered installations."""
+    # it doesn't do anything for a named run
+    tmp_run_dir('foo', named=True)
```

Comment:
In the `tmp_run_dir` fixture, the `named` arg doesn't do anything unless `installed=True` too. E.g. `tmp_run_dir('foo/bar', installed=True, named=True)`. But in this case I think a simple
```suggestion
    tmp_run_dir('foo/bar')
```
is what you are after

Reply:
Not quite, I wanted to ensure it treats a named run the same way as a hierarchical workflow.

Reply:
Changed.